src/main/java/io/confluent/kafka/connect/datagen/DatagenConnector.java

```diff
@@ -107,4 +137,90 @@ private void validateSchemaSource(Config config) {
           .forEach(v -> v.addErrorMessage(SCHEMA_SOURCE_ERR));
     }
   }
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  private void validateSchema(Config config) {
+    List<ConfigValue> schemaSources = config.configValues().stream()
+            .filter(v -> DatagenConnectorConfig.isExplicitlySetSchemaSource(v.name(), v.value()))
+            .collect(Collectors.toList());
+
+    Generator.Builder generatorBuilder = new Generator.Builder();
+    // don't perform this validation if multiple schema sources provided
+    if (schemaSources.size() != 1) {
+      return;
+    }
+
+    ConfigValue schemaSource = schemaSources.get(0);
+
+    switch (schemaSource.name()) {
+      case DatagenConnectorConfig.QUICKSTART_CONF:
+        if (!Quickstart.configValues.contains(((String) schemaSource.value()).toLowerCase())) {
+          schemaSource.addErrorMessage(QUICKSTART_ERR);
+          break;
+        }
+        try {
+          generator = generatorBuilder
+                  .schemaStream(getClass().getClassLoader()
+                          .getResourceAsStream(Quickstart.valueOf(((String) schemaSource.value())
+                                  .toUpperCase()).getSchemaFilename()))
+                  .build();
+        } catch (SchemaParseException | IOException e) {
+          schemaSource.addErrorMessage(QUICKSTART_SCHEMA_ERR);
+        }
+        break;
+
+      case DatagenConnectorConfig.SCHEMA_STRING_CONF:
+        try {
+          generator = generatorBuilder.schemaString((String) schemaSource.value()).build();
+        } catch (SchemaParseException e) {
+          schemaSource.addErrorMessage(SCHEMA_STRING_ERR);
+        }
+        break;
+
+      case DatagenConnectorConfig.SCHEMA_FILENAME_CONF:
+        try {
+          InputStream stream = new FileInputStream((String) schemaSource.value());
+          generator = generatorBuilder.schemaStream(stream).build();
+        } catch (FileNotFoundException e) {
+          try {
+            if (DatagenTask.class.getClassLoader()
+                    .getResource((String) schemaSource.value()) == null) {
+              schemaSource.addErrorMessage(SCHEMA_FILE_NOT_FOUND);
+              break;
+            }
+            generator = generatorBuilder.schemaStream(
+                    DatagenConnector.class.getClassLoader()
+                            .getResourceAsStream((String) schemaSource.value())).build();
+          } catch (SchemaParseException | IOException i) {
+            schemaSource.addErrorMessage(SCHEMA_FILE_ERR);
+          }
+        } catch (SchemaParseException | IOException e) {
+          schemaSource.addErrorMessage(SCHEMA_FILE_ERR);
+        }
+        break;
+
+      default:
+        break;
+    }
+  }
+
+  private void validateSchemaKeyField(Config config) {
+    // don't perform this validation if provided schema is invalid
+    if (generator == null) {
+      return;
+    }
```

Comment:
`Connector` objects are reused by the framework for preflight validations and a single one may even be used simultaneously. We should avoid tracking validation state via instance variables. Instead, we might consider adding a `boolean` return value from preceding validation methods that indicates whether they succeeded or not and, depending on that return value, decide in the `validate` method whether to call follow-up validation methods.

Reply:
Thanks, that's valuable feedback. I was tracking the state since the parsed schema was being reused in the `validateSchemaKeyField` method. What I'm doing now is returning a Schema from the preceding validation which is null in case of invalid schema and passing the schema to the next validation